File: Segundo/Segundo_Cuatrimestre/Informatica/PDOO/SUBIR/MARCIANITOS_P3/Marcianitos/src/deepspace/EnemyStarShip.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

/**
 *
 * @author laura
 */
public class EnemyStarShip {    // funciona
    private float ammoPower;
    private String name;
    private float shieldPower;
    private Loot loot;
    private Damage damage;
    
    EnemyStarShip(String n, float a, float s, Loot l, Damage d) {
        this.name = n;
        this.ammoPower = a;
        this.shieldPower = s;
        this.loot = l;
        this.damage = d;
    }
    
    EnemyStarShip(EnemyStarShip e) {
        this.name = e.name;
        this.ammoPower = e.ammoPower;
        this.shieldPower = e.shieldPower;
        this.loot = e.loot;
        this.damage = e.damage;
    }
    
    EnemyToUI getUIversion() {
        return new EnemyToUI(this);
    }
    
    public float fire(){
        return this.getAmmoPower();
    }
    
    public float getAmmoPower(){
        return this.ammoPower;
    }
    
    public Damage getDamage(){
        return this.damage;
    }
    
    public Loot getLoot(){
        return this.loot;
    }
    
    public String getName(){
        return this.name;
    }
    
    public float getShieldPower(){
        return this.shieldPower;
    }
    
    public float protection(){
        return this.getShieldPower();
    }
    
    public ShotResult receiveShot(float shot){
        ShotResult resultado;
        
        if (protection() < shot)
            resultado = ShotResult.DONOTRESIST;
        else
            resultado = ShotResult.RESIST;
       
        return resultado;           
    }
}

File: Segundo/Segundo_Cuatrimestre/Informatica/PDOO/SUBIR/Examen19-20/EnemyStarShip.rb
#encoding: UTF-8

require_relative 'ShotResult'
require_relative 'Loot'
require_relative 'Damage'

require_relative 'EnemyToUI'

#require_relative 'SuppliesPackage'

module Deepspace
	class EnemyStarShip
		# Constructor
		def initialize (n, ap, sp, l, d)  # n string, ap float, sp float, l loot, d Damage
			@name = n
			@ammoPower = ap
			@shieldPower = sp
			@loot = l
			@damage = d
			
		end
		
		# Constructor de copia
		def self.newCopy(other)
			new(other.name, other.ammoPower, other.shieldPower, other.loot, other.damage)
		end
		
		def getUIversion()
			return EnemyToUI.new(self)
		end
		
		# Consultores publicos
		attr_reader :ammoPower
		attr_reader :loot
		attr_reader :name
		attr_reader :shieldPower
		attr_reader :damage
		
		def protection ()
			return @shieldPower   # devuelve float
		end
		
		def fire ()
			return @ammoPower   # devuelve float
		end
		
		def receiveShot (shot)   # shot es un float
			resultado = ShotResult::RESIST
			
			if (protection() < shot)
				resultado = ShotResult::DONOTRESIST
			end
			
			return (resultado)   # devuelve un ShotResult
		end
		
		def to_s
			out = "ENEMY - Name: " + @name + ", ammoPower:  #{@ammoPower}, shieldPower: #{@shieldPower} \n"
			out += "\tLoot: " + @loot.to_s + "\n"
			out += "\tDamage: " + @damage.to_s + "\n"
			out += "------- end of Enemy " + @name + " -------"
			return out
		end
		
	end	#class

	#d = Damage.newNumericWeapons(3, 2)
	#puts d.getUIversion

	#l = Loot.new(SuppliesPackage.new(1,2,3), 2, 3, 1, 5)
	#l.getUIversion

	#e = EnemyStarShip.new("Huebo", 2, 3, l, d)
	#puts e.to_s
	
	#eui = EnemyToUI.new(e)
	#puts eui.to_s

end	#module
